packages/scoutgame/src/projects/createScoutProject.ts

```diff
@@ -1,11 +1,84 @@
+import { log } from '@charmverse/core/log';
 import type { ScoutProjectMemberRole } from '@charmverse/core/prisma-client';
 import { prisma } from '@charmverse/core/prisma-client';
+import {
+  uploadUrlToS3,
+  getFilenameWithExtension,
+  uploadFileToS3,
+  getUserS3FilePath
+} from '@packages/aws/uploadToS3Server';
+import { getContractDeployerAddress } from '@packages/blockchain/getContractDeployerAddress';
+import sharp from 'sharp';
+import { verifyMessage } from 'viem';
 
+import { CONTRACT_DEPLOYER_SIGN_MESSAGE } from './constants';
 import type { CreateScoutProjectFormValues } from './createScoutProjectSchema';
 import { generateProjectPath } from './generateProjectPath';
+import { generateRandomAvatar } from './generateRandomAvatar';
 
 export async function createScoutProject(payload: CreateScoutProjectFormValues, userId: string) {
   const path = await generateProjectPath(payload.name);
+  const contractTransactionRecord: Record<string, { txHash: string; blockNumber: number; blockTimestamp: number }> = {};
+
+  if (payload.deployers) {
+    for (const deployer of payload.deployers) {
+      const isValidSignature = await verifyMessage({
+        message: CONTRACT_DEPLOYER_SIGN_MESSAGE,
+        signature: deployer.signature as `0x${string}`,
+        address: deployer.address as `0x${string}`
+      });
+
+      if (!isValidSignature) {
+        throw new Error(`Invalid signature for deployer ${deployer.address}`);
+      }
+    }
+  }
+
+  if (payload.contracts) {
+    for (const contract of payload.contracts) {
+      const { block, transaction } = await getContractDeployerAddress({
+        contractAddress: contract.address.toLowerCase(),
+        chainId: contract.chainId
+      });
+
+      contractTransactionRecord[contract.address.toLowerCase()] = {
+        txHash: transaction.hash,
+        blockNumber: Number(block.number),
+        blockTimestamp: Number(block.timestamp)
+      };
+
+      if (contract.deployerAddress.toLowerCase() !== transaction.from.toLowerCase()) {
+        throw new Error(
```

Comment:
hm maybe we should just ignore the deployer provided by the frontend, if you're looking up the correct one anyway